File: residential_controller.py
class Column: 
    def __init__ (self, elevatorNumbers, numberFloors):
        self.elevatorList = []
        self.floorsList = []

#           ***** CREATING FLOORLIST *****
        for i in range(numberFloors):
            self.floorsList.append(i+1)

#           ***** CREATING ELEVATORLIST WITH OBJECT ELEVATOR *****
        for i in range(elevatorNumbers):
            elevator = Elevator(1, "", "elevator " + str(i+1))
            self.elevatorList.append(elevator)

    
    def requestElevator(self, requestedFloor, requestDirection):
        self.requestedFloor = requestedFloor
        self.requestDirection = requestDirection
        self.findElevator(requestedFloor, requestDirection)

    def requestFloor(self, destination):
        self.destination = destination
        self.elevatorList[0].moveD(destination)
        

#              ***** FIND BEST ELEVATOR FOR SCENARIO DEPENDING ON VARIABLES *****
    def findElevator(self, requestedFloor, requestDirection):
        for elevator in self.elevatorList:
            floorsDiff = abs(elevator.currentFloor - requestedFloor)
            if(elevator.currentFloor == requestedFloor and elevator.direction =="idle"):
                elevator.elevatorScore = 0 + floorsDiff
            elif(elevator.currentFloor < requestedFloor and requestDirection == elevator.direction and elevator.direction == "up"):
                elevator.elevatorScore = 100 + floorsDiff
            elif(elevator.currentFloor > requestedFloor and requestDirection == elevator.direction and elevator.direction == "down"):
                elevator.elevatorScore = 200 + floorsDiff
            elif(elevator.currentFloor == requestedFloor and elevator.direction == "idle"):
                elevator.elevatorScore = 300 + floorsDiff
            elif(elevator.currentFloor != requestedFloor and elevator.direction != requestDirection and elevator.direction != "idle"):
                elevator.elevatorScore = 400 + floorsDiff
            elif(elevator.currentFloor == requestedFloor and elevator.direction != requestDirection):
                elevator.elevatorScore = 500 + floorsDiff
            elif(elevator.currentFloor != requestedFloor and elevator.direction == requestDirection):
                elevator.elevatorScore = 600 + floorsDiff
            elif(elevator.direction == "idle"):
                elevator.elevatorScore = 700 + floorsDiff
        self.elevatorList.sort(key=lambda x: x.elevatorScore)
        bestElevator = self.elevatorList[0]
        bestElevator.move(requestedFloor)


class Elevator:
    def __init__ (self, currentFloor, direction, name):
        self.requestFloor = []
        self.direction = direction
        self.currentFloor = currentFloor
        self.name = name
        self.elevatorScore = 1000

    def move(self, requestFloor):
        print(self.name, "is sent.", self.name, " is currently at floor : ", self.currentFloor)
        while(self.currentFloor < requestFloor):
            self.direction = "up"
            self.currentFloor += 1
            print(self.name , "at floor : " , self.currentFloor)
        while(self.currentFloor > requestFloor):
            self.direction = "down"
            self.currentFloor -= 1
            print(self.name , "at floor: " , self.currentFloor)
        print("============================================================")
        print(self.name, "is arrived at destination. Doors open. You enter.")
        print("============================================================")
        self.direction = "idle"
        

    def moveD(self, destination):
        while(self.currentFloor < destination):
            self.direction = "up"
            self.currentFloor += 1
            print(self.name , "at floor : " , self.currentFloor)
        while(self.currentFloor > destination):
            self.direction = "down"
            self.currentFloor -= 1
            print(self.name , "at floor: " , self.currentFloor)
        print("===========================================================")
        print(self.name, " is arrived. Doors open. You exit.")
        print("===========================================================")
        self.direction = "idle"



    
#                                   ***** Changeable Variables *****
#If you change elevatorNumbers, make sure to uncomment "column1.elevatorList[i].blabla" below

#requestDirection = "up"
#requestFloor = 5
#destination = 10


#column1 = Column(elevatorNumbers, numberFloors)

#                   ***** SCENARIOS *****

#                   ***** SCENARIO #1 *****
def scenario():
    while True:
        try:
            scenario = int(input("Which scenario you'd like to start with? (Type 1, 2 or 3) :"))
            while(scenario < 1 or scenario > 3):
                scenario = int(input("Please enter a valid number."))

            if(scenario == 1):
                #Elevator 1 idle at floor 2, elevator 2 idle at floor 6, someone request elevator at floor 3 and wants to go to Floor 7
                elevatorNumbers = 2 
                numberFloors = 10
                column1 = Column(elevatorNumbers, numberFloors)
                #RequestElevator sent at floor 3
                requestFloor = 3
                requestDirection = "up"
                #Wants to go to floor 7
                destination = 7
                #elevator 1
                column1.elevatorList[0].currentFloor = 2
                column1.elevatorList[0].direction = "idle"
                #elevator 2
                column1.elevatorList[1].currentFloor = 6
                column1.elevatorList[1].direction = "idle"
                #Move elevator to floor 3, then floor 7
                column1.requestElevator(requestFloor, requestDirection)
                column1.requestFloor(destination)

                retry()

            elif(scenario == 2):
                #Elevator 1 idle at floor 10, elevator 2 idle at floor 3, someone on 1st floor and request 6th floor.
                elevatorNumbers = 2 
                numberFloors = 10
                column1 = Column(elevatorNumbers, numberFloors)
                #RequestElevator sent at floor 1
                requestFloor = 1
                requestDirection = "up"
                #Wants to go to floor 6
                destination = 6
                #elevator 1 idle at floor 10
                column1.elevatorList[0].currentFloor = 10
                column1.elevatorList[0].direction = "idle"
                #elevator 2 idle at floor 3
                column1.elevatorList[1].currentFloor = 3
                column1.elevatorList[1].direction = "idle"
                #Move elevator to floor 1, then floor 6
                column1.requestElevator(requestFloor, requestDirection)
                column1.requestFloor(destination)

                print("")
                print("*****2 minutes later, another request is sent*****")
                print("")

                #New request from floor 3
                requestFloor = 3
                requestDirection = "up"
                #Wants to go to floor 5
                destination = 5
                #move elevator to floor 3, then 5
                column1.requestElevator(requestFloor, requestDirection)
                column1.requestFloor(destination)

                print("")
                print("*****A third request is sent.*****")
                print("")

                #New request from floor 3
                requestFloor = 9
                requestDirection = "down"
                #Wants to go to floor 2
                destination = 2
                #move elevator to floor 9, then 2
                column1.requestElevator(requestFloor, requestDirection)
                column1.requestFloor(destination)

                retry()
            
            elif(scenario == 3):
                #Elevator 1 idle at floor 10, elevator 2 moving up from floor 3 to 6, someone on 3rd floor and request 2nd floor.
                elevatorNumbers = 2 
                numberFloors = 10
                column1 = Column(elevatorNumbers, numberFloors)
                #RequestElevator sent at floor 3
                requestFloor = 3
                requestDirection = "down"
                #Wants to go to floor 6
                destination = 2
                #elevator 1 idle at floor 10
                column1.elevatorList[0].currentFloor = 10
                column1.elevatorList[0].direction = "idle"
                #elevator 2 moving up at floor 3 to 6
                column1.elevatorList[1].currentFloor = 3
                column1.elevatorList[1].direction = "up"
                #Move elevator to floor 1, then floor 6
                column1.requestElevator(requestFloor, requestDirection)
                column1.requestFloor(destination)

                print("")
                print("*****5 minutes later, another request is sent.*****")
                print("")

                #New request from floor 10
                requestFloor = 10
                requestDirection = "down"
                #Wants to go to floor 3
                destination = 3
                #elevator 2 is now at floor 6
                column1.elevatorList[1].currentFloor = 6
                column1.elevatorList[1].direction = "idle"
                #move elevator to floor 3, then 5
                column1.requestElevator(requestFloor, requestDirection)
                column1.requestFloor(destination)
                retry()
            break
        except ValueError:
            print("Please enter a valid number.")

def retry():
    retry = input("Want to retry another scenario? y/n : ")
    while(retry != "y" and retry != "n"):
        input("please choose y or n.").lower
        print(retry)
    if(retry == "y"):
        scenario()
    else:
        print("Thank you for trying out my program. Exiting.")

scenario()

File: residential_Controller.js
/*                            ***** RUN JS IN https://repl.it/@SamuelJubinvill/residentialController   ******
                                                  Also, sorry for the inputs ;) 
*/

class Column {
  constructor(numberFloors, numberElevators){
      this.elevatorList = [];
      this.floorsList = [];

//                               ***** SETTING FLOORLIST *****
      for(let i = 1; i <= numberFloors ; i++){
          this.floorsList.push(i);
      }
//        ***** SETTING ELEVATORLIST AND ELEVATORS DIRECTION AND ELEVATORS POSITION *****
      for(var i = 1 ; i <= numberElevators; i++){
          var currentFloor = parseInt(prompt(" Where is Elevator " + [i] + "? floor :", ""));
            while(isNaN(currentFloor) || currentFloor < 1 || currentFloor > numberFloors){
              currentFloor = parseInt(prompt(" please enter a valid number for elevator's floor"));
            }
            var direction = String(prompt("Where is it heading", "")).toLowerCase();
            while (direction != "up" && direction != "down" && direction != "null" && direction != "idle"){
              direction = String(prompt("Please choose between up, down or null/idle"));
            }
            if(direction == "idle"){
              direction = "null";
            }
            let id = "Elevator " + parseInt(i) ;
            var elevator = new Elevator(currentFloor, direction, id);
            this.elevatorList.push(elevator);
            }
      }
  
//                    ***** REQUEST ELEVATOR FROM USER FLOOR *****

requestElevator(requestFloor, requestDirection){
  var requestFloor = parseInt(prompt("What floor are you requesting the elevator?"));
  while(isNaN(requestFloor) || requestFloor < 1 || requestFloor > numberFloors){
  requestFloor = parseInt(prompt("Please enter a valid number for request floor"));
  }
var requestDirection = String(prompt("Going up or down?")).toLowerCase();
  while (requestDirection != "up" && requestDirection != "down" ){
    requestDirection = String(prompt("Please choose between up or down values")).toLowerCase();
  }
column1.findElevator(requestFloor, requestDirection);
}

//                            ***** FIND ELEVATOR *****

findElevator(requestFloor, requestDirection){
this.elevatorList.forEach(function(elevator){
  if(elevator.currentFloor == requestFloor && requestDirection == elevator.direction){
    elevator.elevatorScore = 0;
  }
  else if(elevator.currentFloor < requestFloor && requestDirection == elevator.direction && elevator.direction == "up"){
    elevator.elevatorScore = 10;
  }
  else if(elevator.currentFloor > requestFloor && requestDirection == "down" && elevator.direction == requestDirection){
    elevator.elevatorScore = 20;
  }
  else if(elevator.currentFloor == requestFloor && elevator.direction == "null"){
    elevator.elevatorScore = 30;
  }
  else if(elevator.direction == "null"){
    elevator.elevatorScore = 40;
  }
  else if(elevator.currentFloor == requestFloor && elevator.direction != requestDirection){
    elevator.elevatorScore = 50;
  }
  else if(elevator.currentFloor != requestFloor && elevator.direction == requestDirection){
    elevator.elevatorScore = 60;
  }
  else if(elevator.currentFloor != requestFloor && elevator.direction != requestDirection && elevator.direction != "null"){
    elevator.elevatorScore = 70;
  }
  console.log(elevator.name + ' score is :' + elevator.elevatorScore);
})
  column1.elevatorList.sort(compare);
  var bestElevator = column1.elevatorList[0];
  bestElevator.move(requestFloor);

  function compare(a, b) {
    if(a.elevatorScore == b.elevatorScore){
     console.log("Recalculating score depending on elevators' position...");
     column1.elevatorList.forEach(function(elevator){
       var floorsDiff = Math.abs(elevator.currentFloor - requestFloor);
       elevator.elevatorScore = elevator.elevatorScore + floorsDiff;
       console.log(elevator.name + ' score is now :' + elevator.elevatorScore);
     });
   }

   
   return a.elevatorScore - b.elevatorScore;
   
}  
}

//                *****FUNCTION TO FIND DESTINATION*****

destination(requestFloor){
  requestFloor = parseInt(prompt("What floor are you going?"));
  while (isNaN(requestFloor) || requestFloor < 0 || requestFloor > numberFloors){
    requestFloor = parseInt(prompt("please enter a valid number."));
  }
  var elevator = this.elevatorList[0];
  elevator.moveD(requestFloor);
}

}

class Elevator{
  constructor(currentFloor, direction, id, elevatorScore){
      this.requestFloor = [];
      this.direction = direction;
      this.currentFloor = currentFloor;
      this.name = id;
      this.elevatorScore = elevatorScore;
      
      }

//                         ***** MOVE ELEVATOR TO USER *****
  move(requestFloor){
    console.log(this.name + " Is sent. " + this.name + " Is currently at floor : " + this.currentFloor);
    while(this.currentFloor < requestFloor){
      this.direction = "up";
      this.currentFloor++; 
      console.log(this.name + " is at floor: " + this.currentFloor);
    }
    while(this.currentFloor > requestFloor){
      this.direction = "down";
      this.currentFloor--; 
      console.log(this.name + " at floor: " + this.currentFloor);
    }
    console.log("==============================================");
    console.log(this.name +" is arrived. Doors open. You enter.");
    console.log("Doors closed.")
    this.direction = "null";
    column1.destination(requestFloor);
  }

//                  ***** MOVE ELEVATOR TO DESTINATION *****

  moveD(requestFloor){
    while(this.currentFloor < requestFloor){
      this.direction = "up";
      this.currentFloor++; 
      console.log(this.name + " at floor: " + this.currentFloor);
    }
    while(this.currentFloor > requestFloor){
      this.direction = "down";
      this.currentFloor--; 
      console.log(this.name + " at floor: " + this.currentFloor);
    }
    console.log("=============================================");
    console.log(this.name +" is arrived. Doors open. You exit.");
    this.direction = "null";
    retry();
}
}

//               ***** SETTING SITUATION PARAMETERS *****

var numberElevators = parseInt(prompt("How many elevators?"));
while (isNaN(numberElevators) || numberElevators < 1) {
numberElevators = prompt("please enter a valid number of elevator", "");}
numberFloors = prompt("How many floors?");
while(isNaN(numberFloors) || numberFloors < 1){
numberFloors = parseInt(prompt("please enter a valid number of Floors", ""));
}

//              ***** RETRY PROGRAM WITH SAME PARAMETERS *****
function retry(){
var accept = String(prompt("Do you want to continue running this programm with same values? y/n")).toLowerCase();
if(accept == "y"){
  this.elevatorScore=('');
  column1.requestFloor();
}
if(accept == "n"){
  console.log("exiting program. Thank you.");
} else {
  accept = String(prompt("Please type 'y' for yes or 'n' for no."));
}
}

let column1 = new Column(numberFloors, numberElevators);

column1.requestElevator();